### PR TITLE
Handle pre_spawn_start possible exceptions

### PR DESCRIPTION
**Pull request by GeorgianaElena**
This should fix https://github.com/jupyterhub/jupyterhub/issues/2683 :eyes:

### CODE REVIEW

File: jupyterhub/handlers/pages.py [hunk 1]
```diff
@@ -172,7 +172,19 @@ def _render_form(self, for_user, spawner_options_form, message=''):
...
+            if f in done:
+                future, = done
+                exc = future.exception()
+                if exc:
+                    raise web.HTTPError(
+                        500,
+                        "Error in Authenticator.pre_spawn_start: %s %s"
+                        % (type(exc).__name__, str(exc)),
+                    )
+                    return
```
Comment (minrk) [inline jupyterhub/handlers/pages.py]: No need for return after raise
```suggestion
```

File: jupyterhub/handlers/pages.py [hunk 2]
```diff
@@ -172,7 +172,19 @@ def _render_form(self, for_user, spawner_options_form, message=''):
                 spawner._spawn_future = None
             # not running, no form. Trigger spawn and redirect back to /user/:name
             f = asyncio.ensure_future(self.spawn_single_user(user, server_name))
-            await asyncio.wait([f], timeout=1)
+            done, pending = await asyncio.wait([f], timeout=1)
+            # If spawn_single_user throws an exception, raise a 500 error
+            # otherwise it may cause a redirect loop
+            if f in done:
```
Comment (minrk) [inline jupyterhub/handlers/pages.py]: We only need to take this branch if f is done and failed:

```suggestion
            if f.done() and f.exception():
```

File: jupyterhub/handlers/pages.py [hunk 3]
```diff
@@ -172,7 +172,19 @@ def _render_form(self, for_user, spawner_options_form, message=''):
...
+                future, = done
+                exc = future.exception()
```
Comment (minrk) [inline jupyterhub/handlers/pages.py]: No need for separate `future` referring to the same `f`

**Comment by GeorgianaElena** [comment]
Addressed the feedback! Thanks a lot @minrk.

File: jupyterhub/handlers/pages.py [hunk 4]
```diff
@@ -172,7 +172,17 @@ def _render_form(self, for_user, spawner_options_form, message=''):
                 spawner._spawn_future = None
             # not running, no form. Trigger spawn and redirect back to /user/:name
             f = asyncio.ensure_future(self.spawn_single_user(user, server_name))
-            await asyncio.wait([f], timeout=1)
+            done, pending = await asyncio.wait([f], timeout=1)
+            # If spawn_single_user throws an exception, raise a 500 error
+            # otherwise it may cause a redirect loop
+            if f.done() and f.exception():
+                exc = f.exception()
+                if exc:
```
Comment (minrk) [inline jupyterhub/handlers/pages.py]: This `if exc` is no longer needed, since we know it's true from above.

Comment (GeorgianaElena) [reply jupyterhub/handlers/pages.py]: ups... sorry, I missed it :disappointed:

File: jupyterhub/handlers/pages.py [hunk 5]
```diff
@@ -172,7 +172,16 @@ def _render_form(self, for_user, spawner_options_form, message=''):
                 spawner._spawn_future = None
             # not running, no form. Trigger spawn and redirect back to /user/:name
             f = asyncio.ensure_future(self.spawn_single_user(user, server_name))
-            await asyncio.wait([f], timeout=1)
+            done, pending = await asyncio.wait([f], timeout=1)
```
Comment (aar0nTw) [inline jupyterhub/handlers/pages.py]: I thought [line 216](https://github.com/jupyterhub/jupyterhub/pull/2684/files#diff-f14530788c189c34e88032abd7c584e0R216) in post handler need rewrite in same way as well.

**Pull request merged**